src/database.ts

```diff
@@ -0,0 +1,100 @@
+import { type Result } from "./registry/index.js";
+import {
+  type Config,
+  type AutoWaitConfig,
+  extractBaseUrl,
+  type ChainName,
+  getBaseUrl,
+  type Signal,
+  type Signer,
+  normalize,
+} from "./helpers/index.js";
+import { Statement } from "./statement.js";
+
+export class Database<D = unknown> {
+  readonly config: Config & Partial<AutoWaitConfig>;
+
+  constructor(config: Config & Partial<AutoWaitConfig> = {}) {
+    this.config = config;
+  }
+
+  static async readOnly(chainNameOrId: ChainName | number): Promise<Database> {
+    const baseUrl = getBaseUrl(chainNameOrId);
+    return new Database({ baseUrl });
+  }
+
+  static async forSigner(signer: Signer): Promise<Database> {
+    const baseUrl = await extractBaseUrl({ signer });
+    return new Database({ signer, baseUrl });
+  }
+
+  prepare<T = D>(sql: string): Statement<T> {
+    return new Statement<T>(this.config, sql);
+  }
+
+  /**
+   * @param statements
+   * @returns
+   */
+  async batch<T = D>(
```

Comment:
On Tabeland, batching statements together requires some finesse. This is the first pass at this. Ideally, we can do more "smart" splitting to handle things like mixing creates, mutations, and reads. Interestingly, the D1 API treats batches as a kinda unique thing, where multiple reads would lead to multiple responses etc.

Note here that writes are done as a single "transaction" (in the smart contract sense), and reads are done "concurrently". This means that writes _have_ to follow the rules of our validator in terms of grouping. You can't mix creates with anything else, you can't mix reads and writes, but you can mix revoke/grant with writes.

Reply:
What do you think of my approach here @joewagner ? Does this seem reasonable for now?

Reply:
Yeah, this looks good for now.  The long outstanding [runSQLs PR](https://github.com/tablelandnetwork/evm-tableland/pull/257) in evm-tableland to enable an array of statements will be pretty relevant to this once its finished.  I'l try to prioritize that and have it rfr by next week. 'Till then this looks good!